// restly/dist/services/Googlesheets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Googlesheets = void 0;
class Googlesheets {
    constructor(httpRequest) {
        this.httpRequest = httpRequest;
    }
    /**
     * Read data from a spreadsheet
     * @returns void
     * @throws ApiError
     */
    read({ sheetId, sheetName, }) {
        return this.httpRequest.request({
            method: 'POST',
            url: '/googlesheets/read',
            formData: {
                'sheetId': sheetId,
                'sheetName': sheetName,
            },
        });
    }
    /**
     * Adds a row to a spreadsheet
     * @returns void
     * @throws ApiError
     */
    addRow({ sheetId, value, sheetName, }) {
        return this.httpRequest.request({
            method: 'POST',
            url: '/googlesheets/addRow',
            formData: {
                'sheetId': sheetId,
                'sheetName': sheetName,
                'value': value,
            },
        });
    }
    /**
     * Adds rows to a spreadsheet
     * @returns void
     * @throws ApiError
     */
    addRows({ sheetId, values, sheetName, }) {
        return this.httpRequest.request({
            method: 'POST',
            url: '/googlesheets/addRows',
            formData: {
                'sheetId': sheetId,
                'sheetName': sheetName,
                'values': values,
            },
        });
    }
}
exports.Googlesheets = Googlesheets;
